heron/statefulstorages/src/java/com/twitter/heron/statefulstorage/hdfs/HDFSStorage.java

```diff
@@ -0,0 +1,198 @@
+    try {
+      in = fileSystem.open(path);
+      state =
+          CheckpointManager.SaveInstanceStateRequest.parseFrom(in);
+    } catch (IOException e) {
+      LOG.log(Level.SEVERE, "Failed to read", e);
+      return false;
+    } finally {
+      SysUtils.closeIgnoringExceptions(in);
+    }
+
+    checkpoint.setCheckpoint(state);
+
+    return true;
+  }
+
+  @Override
+  public boolean dispose(String topologyName, String oldestCheckpointPreserved,
+                         boolean deleteAll) {
+    String topologyCheckpointRoot = getTopologyCheckpointRoot(topologyName);
+    Path topologyRootPath = new Path(topologyCheckpointRoot);
+
+
+    if (deleteAll) {
+      // Clean all checkpoint states
+      try {
+        fileSystem.delete(topologyRootPath, true);
+        if (fileSystem.exists(topologyRootPath)) {
+          return false;
+        }
+      } catch (IOException e) {
+        LOG.log(Level.SEVERE, "Failed to delete: " + topologyCheckpointRoot, e);
+        return false;
+      }
+    } else {
+      try {
+        FileStatus[] statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(oldestCheckpointPreserved) < 0) {
+            fileSystem.delete(status.getPath(), true);
+          }
+        }
+
+        // Do a double check. Now all checkpoints with smaller checkpoint id should be cleaned
+        statuses = fileSystem.listStatus(topologyRootPath);
+        for (FileStatus status : statuses) {
+          String name = status.getPath().getName();
+          if (name.compareTo(oldestCheckpointPreserved) < 0) {
+            return false;
+          }
+        }
+
+      } catch (IOException e) {
+        LOG.log(Level.SEVERE, "Failed to clean to: " + oldestCheckpointPreserved, e);
+        return false;
+      }
+    }
+
+
+    return true;
+  }
+
+  /**
+   * Ensure the existence of a directory.
+   * Will create the directory if it does not exist.
+   *
+   * @param dir The path of dir to ensure existence
+   * @return true if the directory exists after this call.
+   */
+  protected boolean ensureDirExists(String dir) {
+    Path path = new Path(dir);
+
+    try {
+      fileSystem.mkdirs(path);
+      if (!fileSystem.exists(path)) {
+        return false;
+      }
+    } catch (IOException e) {
+      LOG.log(Level.SEVERE, "Failed to mkdirs: " + dir, e);
+      return false;
+    }
+
+    return true;
+  }
+
+  protected String getTopologyCheckpointRoot(String topologyName) {
+    return String.format("%s/%s", checkpointRootPath, topologyName);
+  }
+
+  protected String getCheckpointDir(Checkpoint checkpoint) {
+    return String.format("%s/%s/%s", getTopologyCheckpointRoot(checkpoint.getTopologyName()),
+                         checkpoint.getCheckpointId(), checkpoint.getComponent());
+  }
+
+  protected String getCheckpointPath(Checkpoint checkpoint) {
```

Comment:
Can we call this `getCheckpointFile` so it doesn't sound the same as `getCheckpointDir`?